packages-exp/auth-exp/src/core/auth/auth_impl.test.ts

```diff
@@ -0,0 +1,192 @@
+  initializeAuth
+} from './auth_impl';
+
+use(sinonChai);
+
+const FAKE_APP: FirebaseApp = {
+  name: 'test-app',
+  options: {
+    apiKey: 'api-key',
+    authDomain: 'auth-domain'
+  },
+  automaticDataCollectionEnabled: false,
+  async delete() {}
+};
+
+describe('AuthImpl', () => {
+  let auth: Auth;
+  let persistenceStub: sinon.SinonStubbedInstance<Persistence>;
+
+  beforeEach(() => {
+    persistenceStub = sinon.stub(inMemoryPersistence);
+    auth = initializeAuth(FAKE_APP, { persistence: inMemoryPersistence });
+  });
+
+  afterEach(sinon.restore);
+
+  describe('#updateCurrentUser', () => {
+    it('sets the field on the auth object', async () => {
+      const user = testUser('uid');
+      await auth.updateCurrentUser(user);
+      expect(auth.currentUser).to.eql(user);
+    });
+
+    it('orders async operations correctly', async () => {
+      const users = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(n => {
+        return testUser(`${n}`);
+      });
+
+      persistenceStub.set.callsFake(() => {
+        return new Promise(resolve => {
+          // Force into the async flow to make this test actually meaningful
+          setTimeout(() => resolve(), 1);
+        });
+      });
+
+      await Promise.all(users.map(u => auth.updateCurrentUser(u)));
+      for (let i = 0; i < 10; i++) {
+        expect(persistenceStub.set.getCall(i)).to.have.been.calledWith(
+          sinon.match.any,
+          users[i].toPlainObject()
+        );
+      }
+    });
+
+    it('setting to null triggers a remove call', async () => {
+      await auth.updateCurrentUser(null);
+      expect(persistenceStub.remove).to.have.been.called;
+    });
+  });
+
+  describe('#signOut', () => {
+    it('sets currentUser to null, calls remove', async () => {
+      await auth.updateCurrentUser(testUser('test'));
+      await auth.signOut();
+      expect(persistenceStub.remove).to.have.been.called;
+      expect(auth.currentUser).to.be.null;
+    });
+  });
+
+  describe('#setPersistence', () => {
+    it('swaps underlying persistence', async () => {
+      const newPersistence = browserLocalPersistence;
+      const newStub = sinon.stub(newPersistence);
+      persistenceStub.get.returns(
+        Promise.resolve(testUser('test').toPlainObject())
+      );
+
+      await auth.setPersistence(newPersistence);
+      expect(persistenceStub.get).to.have.been.called;
+      expect(persistenceStub.remove).to.have.been.called;
+      expect(newStub.set).to.have.been.calledWith(
+        sinon.match.any,
+        testUser('test').toPlainObject()
+      );
+    });
+  });
+});
+
+describe('initializeAuth', () => {
+  afterEach(sinon.restore);
+
+  it('throws an API error if key not provided', () => {
+    expect(() =>
+      initializeAuth({
+        ...FAKE_APP,
+        options: {} // apiKey is missing
+      })
+    ).to.throw(
+      FirebaseError,
+      'Firebase: Your API key is invalid]: please check you have copied it correctly. (auth/invalid-api-key).'
```

Comment:
we'll need to either fix the error messages or resolve during merge conflict later

Reply:
Yeah I'll hold off and rebase onto your fix when it merges

Reply:
ready!

Reply:
Already done :)